libmscore/layout.cpp

```diff
@@ -3500,33 +3469,131 @@ System* Score::collectSystem(LayoutContext& lc)
+                        voltas.push_back(sp);
+                  }
+            }
+      processLines(system, voltas, false);
+
+      //
+      // vertical align volta segments
+      //
+      for (int staffIdx = 0; staffIdx < nstaves(); ++staffIdx) {
+            std::vector<SpannerSegment*> voltaSegments;
+            for (SpannerSegment* ss : system->spannerSegments()) {
+                  if (ss->isVoltaSegment() && ss->staffIdx() == staffIdx)
+                        voltaSegments.push_back(ss);
+                  }
+            while (!voltaSegments.empty()) {
+                  // we assume voltas are sorted left to right (by tick values)
+                  qreal y = 0;
+                  int idx = 0;
+                  Volta* prevVolta = 0;
+                  for (SpannerSegment* ss : voltaSegments) {
+                        Volta* volta = toVolta(ss->spanner());
+                        if (prevVolta && prevVolta != volta) {
+                              // check if volta is adjacent to prevVolta
+                              if (prevVolta->tick2() != volta->tick())
+                                    break;
+                              }
+                        y = qMin(y, ss->rypos());
+                        ++idx;
+                        prevVolta = volta;
+                        }
+                  for (int i = 0; i < idx; ++i)
+                        voltaSegments[i]->rypos() = y;
+                  voltaSegments.erase(voltaSegments.begin(), voltaSegments.begin() + idx);
+                  }
+            }
+
+      //-------------------------------------------------------------
+      // FretDiagram
+      //-------------------------------------------------------------
+
+      if (hasFretDiagram) {
+            for (const Segment* s : sl) {
+                  for (Element* e : s->annotations()) {
+                        if (e->isFretDiagram())
+                              e->layout();
+                        }
+                  }
+
+            //-------------------------------------------------------------
+            // Harmony, 2nd place
+            // We have FretDiagrams, we want the Harmony above this and
+            // above the volta.
+            //-------------------------------------------------------------
+
+            for (const Segment* s : sl) {
+                  for (Element* e : s->annotations()) {
+                        if (e->isHarmony()) {
+                              Harmony* h = toHarmony(e);
+                              // Layout of harmony seems to be bound currently
+                              // to ChordRest (see ChordRest::shape). But harmony
+                              // can exist without chord or rest too.
+                              if (h->isLayoutInvalid())
+                                    h->layout();
+                              toHarmony(e)->autoplaceSegmentElement(styleP(Sid::minHarmonyDistance));
+                              }
```

Comment:
Wound it perhaps be better to extract this to a separate function? It seems this chunk of code is just repeated twice, so any possible changes to them should be repeated which is easy to forget.